[1.1.x] Support some more LCD types

jmdearras opened the pull request:
Added 4 displays

jmdearras committed:
add 2 displays

original CR-10 and MKS Minidisplay

jmdearras committed:
"Standard" Display

jmdearras committed:
Edit CR10_stockdisplay NOPs

jmdearras committed:
Added Displays cleanup

jmdearras committed:
Added 2 line display

jmdearras committed:
More cleanup

Roxy-3D committed:
The default board needs to be RAMPS_EFB

Roxy-3D commented:
All of the example/configuration.h files need to be changed also.

fiveangle commented on Marlin/Conditionals_LCD.h:
```diff
@@ -108,7 +108,26 @@
     #define NEWPANEL
 
-  #endif
+  #elif ENABLED(CR10_stockdisplay)
```
Since these are just delays specific to this board when using that display, shouldn't these be (and already are in) pins_MELZI_CREALITY.h ?

https://github.com/MarlinFirmware/Marlin/blob/1.1.x/Marlin/pins_MELZI_CREALITY.h#L60-L63

fiveangle replied:
I was looking at something else and saw this, so I think your REVERSE_MENU_DIRECTION might over-rided by Configuration.h without including any ifdef/ifenabled checks (although it mentions only `pins.h` and `Configuration_adv.h`):

```
/**
 * Conditionals_LCD.h
 * Conditionals that need to be set before Configuration_adv.h or pins.h
 */

#ifndef CONDITIONALS_LCD_H // Get the LCD defines which are needed first
```

To be 100% sure, just put your changes into Conditionals_LCD.h, then compile / load on your printer with and without REVERSE_MENU_DIRECTION set in Configuration.h and ensure it changes.  If it does, you should be good so no changes would be needed.

fiveangle replied:
See my comment on next code block re: define precedence, which may make this okay to keep as well.

jmdearras replied:
I removed the menu direction

fiveangle commented on Marlin/Conditionals_LCD.h:
```diff
@@ -108,7 +108,26 @@
+    #define U8GLIB_SH1106 
+    #define REPRAP_DISCOUNT_SMART_CONTROLLER 
+    #define NEWPANEL 
+    #define REVERSE_MENU_DIRECTION
```
Won't this override in all circumstances for this display ?  Doesn't this need to honor REVERSE_MENU_DIRECTION if set in Configuration.h ?

fiveangle replied:
I was looking at something else and saw this, so I think your `REVERSE_MENU_DIRECTION` might over-rided by `Configuration.h` without including any ifdef/ifenabled checks (although it mentions only `pins.h` and `Configuration_adv.h`):

```
/**
 * Conditionals_LCD.h
 * Conditionals that need to be set before Configuration_adv.h or pins.h
 */

#ifndef CONDITIONALS_LCD_H // Get the LCD defines which are needed first
```

To be 100% sure, just put your changes into `Conditionals_LCD.h`, then compile / load on your printer with and without `REVERSE_MENU_DIRECTION` set in `Configuration.h` and ensure it changes.  If it does, you should be good so no changes would be needed.

fiveangle commented on Marlin/ultralcd_impl_DOGM.h:
```diff
@@ -180,9 +180,13 @@
   // Generic support for SH1106 OLED I2C LCDs
   //U8GLIB_SH1106_128X64 u8g(U8G_I2C_OPT_NONE | U8G_I2C_OPT_FAST);  // 8 stripes
-  U8GLIB_SH1106_128X64_2X u8g(U8G_I2C_OPT_NONE | U8G_I2C_OPT_FAST); // 4 stripes
+  U8GLIB_SH1106_128X64_2X u8g(U8G_I2C_OPT_NONE | U8G_I2C_OPT_FAST); // 4 stripes //bunnyhack
```
What is "bunnyhack" ? :)

jmdearras replied:
place keeper, sorry, removed!

fiveangle reviewed (commented):
just a few things i was curious about

fiveangle commented on Marlin/Configuration.h:
```diff
@@ -1364,6 +1364,10 @@
+
+//#define MAKEBOARD_MINI_2_LINE_DISPLAY_1602
+
 //
```
Normally new panels are added at the end of the list so that people who are intimately familiar with the Config format can easily find the section start when scanning through it.  Regardless, the entry should also match the comment style:

```
//
// panel name
// link to external info on the panel (ideally reprap.org wiki page)
//
// any notes (optional)
//
//#define PANEL_DEFINE_NAME
```

jmdearras replied:
fixed

jmdearras commented:
This makes the stock display global, works on other boards so CR-10 owners
can upgrade the board and keep the case/display.

Jim

On Mon, Oct 2, 2017 at 10:42 PM, Dave Johnson <notifications@github.com>
wrote:

> *@fiveangle* commented on this pull request.
> ------------------------------
>
> In Marlin/Conditionals_LCD.h
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142301584>
> :
>
> > @@ -108,7 +108,26 @@
>      #define ULTIPANEL
>      #define NEWPANEL
>
> -  #endif
> +  #elif ENABLED(CR10_stockdisplay)
>
> Since these are just delays specific to this board when using that
> display, shouldn't these be (and already are in) pins_MELZI_CREALITY.h ?
>
> https://github.com/MarlinFirmware/Marlin/blob/1.1.x/Marlin/pins_MELZI_
> CREALITY.h#L60-L63
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#pullrequestreview-66642516>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiWZmmXPVeBuBmWQIDGLdOLzvX8cBks5soZ8igaJpZM4PrIQI>
> .
>

jmdearras commented:
I have seen  REVERSE_MENU_DIRECTION coded in other displays, I can remove
it.


On Mon, Oct 2, 2017 at 10:45 PM, Dave Johnson <notifications@github.com>
wrote:

> *@fiveangle* commented on this pull request.
> ------------------------------
>
> In Marlin/Conditionals_LCD.h
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142301856>
> :
>
> > @@ -108,7 +108,26 @@
>      #define ULTIPANEL
>      #define NEWPANEL
>
> -  #endif
> +  #elif ENABLED(CR10_stockdisplay)
> +
> +    #define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
> +    #define ST7920_DELAY_1 DELAY_2_NOP
> +    #define ST7920_DELAY_2 DELAY_2_NOP
> +    #define ST7920_DELAY_3 DELAY_2_NOP
> +
> +  #elif ENABLED (MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
> +
> +    #define U8GLIB_SH1106
> +    #define REPRAP_DISCOUNT_SMART_CONTROLLER
> +    #define NEWPANEL
> +    #define REVERSE_MENU_DIRECTION
>
> Won't this override in all circumstances for this display ? Doesn't this
> need to honor REVERSE_MENU_DIRECTION if set in Configuration.h ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#pullrequestreview-66642809>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiW3sqbLbH684d9Rux9NJ-Qslsfaoks5soZ_VgaJpZM4PrIQI>
> .
>

jmdearras commented:
I am new to making modifications to the marlin code.  I have bought several
boards and displays to use in modifying my printers and made the changes to
get them working, all displays to all boards I have.  Is there a prefered
protocol to submit changes?  I will follow them if there is a reasonable
chance they will be accepted.

Jim De Arras

On Tue, Oct 3, 2017 at 12:20 PM, Jim DeArras <jmdearras@gmail.com> wrote:

> I have seen  REVERSE_MENU_DIRECTION coded in other displays, I can remove
> it.
>
>
> On Mon, Oct 2, 2017 at 10:45 PM, Dave Johnson <notifications@github.com>
> wrote:
>
>> *@fiveangle* commented on this pull request.
>> ------------------------------
>>
>> In Marlin/Conditionals_LCD.h
>> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142301856>
>> :
>>
>> > @@ -108,7 +108,26 @@
>>      #define ULTIPANEL
>>      #define NEWPANEL
>>
>> -  #endif
>> +  #elif ENABLED(CR10_stockdisplay)
>> +
>> +    #define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
>> +    #define ST7920_DELAY_1 DELAY_2_NOP
>> +    #define ST7920_DELAY_2 DELAY_2_NOP
>> +    #define ST7920_DELAY_3 DELAY_2_NOP
>> +
>> +  #elif ENABLED (MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
>> +
>> +    #define U8GLIB_SH1106
>> +    #define REPRAP_DISCOUNT_SMART_CONTROLLER
>> +    #define NEWPANEL
>> +    #define REVERSE_MENU_DIRECTION
>>
>> Won't this override in all circumstances for this display ? Doesn't this
>> need to honor REVERSE_MENU_DIRECTION if set in Configuration.h ?
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/MarlinFirmware/Marlin/pull/7824#pullrequestreview-66642809>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAksiW3sqbLbH684d9Rux9NJ-Qslsfaoks5soZ_VgaJpZM4PrIQI>
>> .
>>
>
>

fiveangle commented:
There is not necessarily a "right" or "wrong" way, but I did pose some comments and questions in your code.  For REVERSE_MENU_DIRECTION, as long as the value in Configuration*.h over-rides the value you added, then it should be fine.  If not, it probably should be left out and the user make the change in Config.  If you think what you have is correct, Scott can review when he merges.

The other things Roxy and I left you didn't respond to (bunnyhack ?  moving new panel to end of list, fixing comment format, updating the rest of the configs, etc).  The protocol is to post your PR like you have, and then respond to comments made (you need to go to the website and respond to the review as you cannot easily do it via email).

jmdearras commented:
can I withdraw the merge, and cleanup the items you and other pointed out,
and then resubmit?
Jim

On Tue, Oct 3, 2017 at 7:12 PM, Dave Johnson <notifications@github.com>
wrote:

> There is not necessarily a "right" or "wrong" way, but I did pose some
> comments and questions in your code. For REVERSE_MENU_DIRECTION, as long as
> the value in Configuration*.h over-rides the value you added, then it
> should be fine. If not, it probably should be left out and the user make
> the change in Config. If you think what you have is correct, Scott can
> review when he merges (but if he agrees that it is wrong, then it may delay
> getting merged if you have to fix it.
>
> The other things Roxy and I left weren't responded to (bunnyhack ? moving
> new panel to end of list, fixing comment format, updating the rest of the
> configs, etc)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#issuecomment-334005473>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiXnXB-CqMb6VDbmFIgcfQpakyDosks5sor93gaJpZM4PrIQI>
> .
>

jmdearras commented:
oh, and do I really need to update all the example configuration.h files?
I do not like to edit what I cannot test well.

On Tue, Oct 3, 2017 at 8:36 PM, Jim DeArras <jmdearras@gmail.com> wrote:

> can I withdraw the merge, and cleanup the items you and other pointed out,
> and then resubmit?
> Jim
>
> On Tue, Oct 3, 2017 at 7:12 PM, Dave Johnson <notifications@github.com>
> wrote:
>
>> There is not necessarily a "right" or "wrong" way, but I did pose some
>> comments and questions in your code. For REVERSE_MENU_DIRECTION, as long as
>> the value in Configuration*.h over-rides the value you added, then it
>> should be fine. If not, it probably should be left out and the user make
>> the change in Config. If you think what you have is correct, Scott can
>> review when he merges (but if he agrees that it is wrong, then it may delay
>> getting merged if you have to fix it.
>>
>> The other things Roxy and I left weren't responded to (bunnyhack ? moving
>> new panel to end of list, fixing comment format, updating the rest of the
>> configs, etc)
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/MarlinFirmware/Marlin/pull/7824#issuecomment-334005473>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAksiXnXB-CqMb6VDbmFIgcfQpakyDosks5sor93gaJpZM4PrIQI>
>> .
>>
>
>

Tannoo commented:
Once you have the config files the way you want... you do have to update all of the example config files to match.

fiveangle commented:
I suggest leave this open and just fix it.

I was looking at something else and noticed that `Conditionals_LCD.h` might load before `Configuration.h` so you may fine there (I commented on that code directly, so take a look).

jmdearras commented:
I hate to sound stupid, but how do I fix it if it's open?

On Wed, Oct 4, 2017 at 8:12 AM, Dave Johnson <notifications@github.com>
wrote:

> I suggest leave this open and just fix it.
>
> I was looking at something else and checked this, so I think your
> REVERSE_MENU_DIRECTION will in fact be over-rided by Configuration*.h
> without including any ifdef/ifenabled checks:
>
> /**
>  * Conditionals_LCD.h
>  * Conditionals that need to be set before Configuration_adv.h or pins.h
>  */
>
> #ifndef CONDITIONALS_LCD_H // Get the LCD defines which are needed first
>
> So disregard that comment from me.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#issuecomment-334135431>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksidxqpDhNSXkgzDzH1dxBjBgRbTL1ks5so3YTgaJpZM4PrIQI>
> .
>

fiveangle commented:
Fix in your local repo then commit to your "Marlin-wip" branch which you have requested the merge for here.  Also, you really should start interacting on the website, as I mentioned... as the email interface is too limiting.  Click the "[view on github](https://github.com/MarlinFirmware/Marlin/pull/7824#issuecomment-334200457)" link, then re-read the info we've put here as a start.

thinkyhead commented:
> I hate to sound stupid, but how do I fix it if it's open?

Any commits you add (and push) to **_your branch_** will show up here.

thinkyhead commented:
@jmdearras I have no problem using the CR-10 example configurations included in the current "nightly" build of Marlin. Are you having issues with them?

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -1146,7 +1146,7 @@ static_assert(1 >= 0
   #if ENABLED(ULTIMAKERCONTROLLER) \
       && DISABLED(SAV_3DGLCD) && DISABLED(miniVIKI) && DISABLED(VIKI2) \
-      && DISABLED(ELB_FULL_GRAPHIC_CONTROLLER) && DISABLED(PANEL_ONE)
+      && DISABLED(ELB_FULL_GRAPHIC_CONTROLLER) && DISABLED(PANEL_ONE) && DISABLED(MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
```
This is not needed, unless enabling `MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER` were to define `ULTIMAKERCONTROLLER` which I don't see that it does.

jmdearras replied:
I got sanity check failure, two displays defined without it.

jmdearras replied:
```cpp
#if defined (MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
 #define DOGLCD
 #define U8GLIB_SH1106
 #define REPRAP_DISCOUNT_SMART_CONTROLLER
 #define NEWPANEL
#endif
```
and
```cpp 
#if ENABLED(PANEL_ONE) || ENABLED(U8GLIB_SH1106)
  #define ULTIMAKERCONTROLLER
#endif
```
it is defined

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -1146,7 +1146,7 @@ static_assert(1 >= 0
+      && DISABLED(ELB_FULL_GRAPHIC_CONTROLLER) && DISABLED(PANEL_ONE) && DISABLED(MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
     + 1
   #endif
   #if ENABLED(REPRAP_DISCOUNT_SMART_CONTROLLER) && DISABLED(REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER)
```
Instead, since enabling `MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER` enables `REPRAP_DISCOUNT_SMART_CONTROLLER`, this is where you would add the clause.

Then, below you would also add:
```cpp
   #if ENABLED(MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
     + 1
   #endif
```

jmdearras replied:
OK, I see.  Done

thinkyhead commented on Marlin/pins_RAMPS.h:
```diff
@@ -129,6 +129,21 @@
+#endif
+#define LCD_PINS_RST  27
+#define LCD_PINS_DC   25
+#endif
```
These pins haven't been defined yet, so there's no need to `#undef` them. Just put your custom values in the LCD block below.

thinkyhead commented on Marlin/pins_RAMPS.h:
```diff
@@ -129,6 +129,21 @@
 #endif
 
+/*---------------MKS OLED patch_4-----------------------*/
+#if defined (MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
```
`#if ENABLED(...)` is required.

thinkyhead commented on Marlin/pins_RAMPS.h:
```diff
@@ -281,9 +314,15 @@
       #define BTN_EN1           31
       #define BTN_EN2           33
       #define BTN_ENC           35
+#endif
```
Indentation counts. Please strive to emulate the indentation style used in the rest of the codebase.

thinkyhead commented on Marlin/pins_RAMPS.h:
```diff
@@ -404,4 +461,4 @@
   #endif // NEWPANEL
 
-#endif // ULTRA_LCD
+ #endif // ULTRA_LCD
```
Remove this space.

thinkyhead commented on Marlin/ultralcd_impl_DOGM.h:
```diff
@@ -180,9 +180,13 @@
   //U8GLIB_SSD1306_128X64 u8g(U8G_I2C_OPT_NONE | U8G_I2C_OPT_FAST);  // 8 stripes
   U8GLIB_SSD1306_128X64_2X u8g(U8G_I2C_OPT_NONE | U8G_I2C_OPT_FAST); // 4 stripes
 #elif ENABLED(U8GLIB_SH1106)
+  #if defined (MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
```
`#if ENABLED(...)`

thinkyhead commented on Marlin/ultralcd_impl_DOGM.h:
```diff
@@ -180,9 +180,13 @@
   U8GLIB_SSD1306_128X64_2X u8g(U8G_I2C_OPT_NONE | U8G_I2C_OPT_FAST); // 4 stripes
 #elif ENABLED(U8GLIB_SH1106)
+  #if defined (MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
+        U8GLIB_SH1106_128X64 u8g(23, 17, 16, 25); // SW SPI Com: SCK = 23, MOSI = 17, CS = 16, A0 = 25
```
Indentation is 2 spaces.

thinkyhead commented on Marlin/Conditionals_LCD.h:
```diff
@@ -108,7 +108,26 @@
+    #define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
+    #define ST7920_DELAY_1 DELAY_2_NOP
+    #define ST7920_DELAY_2 DELAY_2_NOP
+    #define ST7920_DELAY_3 DELAY_2_NOP
```
These must be:
```cpp
    #ifndef ST7920_DELAY_1
      #define ST7920_DELAY_1 DELAY_2_NOP
    #endif
    #ifndef ST7920_DELAY_2
      #define ST7920_DELAY_2 DELAY_2_NOP
    #endif
    #ifndef ST7920_DELAY_3
      #define ST7920_DELAY_3 DELAY_2_NOP
    #endif
```
…but see my note below…

jmdearras replied:
fiveangle had me remove them  Which is it?

jmdearras replied:
I am not following the thread well, apparently.  What note below?

fiveangle replied:
I thought Scott indicated that Conditionals_LCD.h was pre-loaded prior to Config*.h and pins_[board].h so didn't need these checks.  If he says they are needed, so be it.  Sorry for missing this earlier Jim and making more work for you unnecessarily.  Whatever Scott says should over-ride what others say (because if he's wrong, he will be the final one to catch it ;)

jmdearras replied:
ok! :)

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -1485,6 +1494,11 @@
   #define U8GLIB_SH1106
 #endif
 
+//#define CR10_stockdisplay
```
Configuration defines must be `ALL_CAPS`.

thinkyhead replied:
That said… This is just a standard reprap discount graphical display. We don't want to have a separate entry for every machine that happens to have slightly different timing for its display. If you think that special `ST7920_DELAY_#` values are needed, they should be placed in the CR-10 example configurations, and this "faux" display type should be removed.

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -1485,6 +1494,11 @@
+//#define CR10_stockdisplay
+
+// MKS OLED 1.3" 128 × 64 FULL GRAPHICS CONTROLLER 
+#define MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER
```
As mentioned, it's nice to have a link to the display's product page, if possible.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -13561,8 +13561,18 @@ void setup() {
 
+
+  /*---------------MKS OLED patch_3-----------------------*/
+#if defined (MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
```
`#if ENABLED(...)`

thinkyhead replied:
Please remove `/*---------------MKS OLED patch_3-----------------------*/`

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -13561,8 +13561,18 @@ void setup() {
+  pinMode(LCD_PINS_RST, OUTPUT);  
+  digitalWrite(LCD_PINS_RST  , LOW);
+  delay(1000);
+  digitalWrite(LCD_PINS_RST  , HIGH);
```
Please indent, and please use FastIO:
```cpp
  SET_OUTPUT(LCD_PINS_DC);   
  SET_OUTPUT(LCD_PINS_RST);  
  WRITE(LCD_PINS_RST, LOW);
  delay(1000);
  WRITE(LCD_PINS_RST, HIGH);
```

thinkyhead commented:
Thanks for the submission! Several changes are needed. Please review comments.

jmdearras commented:
it is NOT a standard display, only connects to AUX1, different pins.

On Wed, Oct 4, 2017 at 2:15 PM, Scott Lahteine <notifications@github.com>
wrote:

> *@thinkyhead* commented on this pull request.
> ------------------------------
>
> In Marlin/Configuration.h
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142749954>
> :
>
> > @@ -1485,6 +1494,11 @@
>    #define U8GLIB_SH1106
>  #endif
>
> +//#define CR10_stockdisplay
>
> That said… This is just a standard reprap display. We don't want to have a
> separate entry for every machine that happens to have slightly different
> timing for its display. If you think that special ST7920_DELAY_# values
> are needed, they should be placed in the CR-10 example configurations, and
> this should not define a "faux" display type.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142749954>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiceoTwSsC6cmvh63TCnKWvKdpRNqks5so8s7gaJpZM4PrIQI>
> .
>

jmdearras commented:
no, why?


On Wed, Oct 4, 2017 at 1:55 PM, Scott Lahteine <notifications@github.com>
wrote:

> @jmdearras <https://github.com/jmdearras> I have no problem using the
> CR-10 example configurations included in the current "nightly" build of
> Marlin. Are you having issues with them?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#issuecomment-334238165>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiePJiEJuBHkTbShUxMaFoOfnWUajks5so8acgaJpZM4PrIQI>
> .
>

jmdearras commented:
the idea was to make the CR-10 display a general ramps display for CR-10
owners (like me) who want to change the stock control board.  It only
connects to AUX1.

On Wed, Oct 4, 2017 at 2:15 PM, Scott Lahteine <notifications@github.com>
wrote:

> *@thinkyhead* commented on this pull request.
> ------------------------------
>
> In Marlin/Configuration.h
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142749954>
> :
>
> > @@ -1485,6 +1494,11 @@
>    #define U8GLIB_SH1106
>  #endif
>
> +//#define CR10_stockdisplay
>
> That said… This is just a standard reprap display. We don't want to have a
> separate entry for every machine that happens to have slightly different
> timing for its display. If you think that special ST7920_DELAY_# values
> are needed, they should be placed in the CR-10 example configurations, and
> this should not define a "faux" display type.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142749954>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiceoTwSsC6cmvh63TCnKWvKdpRNqks5so8s7gaJpZM4PrIQI>
> .
>

jmdearras commented:
the seller's page?

On Wed, Oct 4, 2017 at 2:19 PM, Scott Lahteine <notifications@github.com>
wrote:

> *@thinkyhead* commented on this pull request.
> ------------------------------
>
> In Marlin/Configuration.h
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142751243>
> :
>
> > @@ -1485,6 +1494,11 @@
>    #define U8GLIB_SH1106
>  #endif
>
> +//#define CR10_stockdisplay
> +
> +// MKS OLED 1.3" 128 × 64 FULL GRAPHICS CONTROLLER
> +#define MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER
>
> As mentioned, it's nice to have a link to the display's product page, if
> possible.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#pullrequestreview-67158880>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksidwxGvE4LG5TWWMdGiypuYiRd15xks5so8wdgaJpZM4PrIQI>
> .
>

jmdearras commented:
what is FastIO and how do I use it?


On Wed, Oct 4, 2017 at 2:23 PM, Scott Lahteine <notifications@github.com>
wrote:

> *@thinkyhead* commented on this pull request.
> ------------------------------
>
> In Marlin/Marlin_main.cpp
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142752501>
> :
>
> >
> +
> +  /*---------------MKS OLED patch_3-----------------------*/
> +#if defined (MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
> +  pinMode(LCD_PINS_DC, OUTPUT);
> +  pinMode(LCD_PINS_RST, OUTPUT);
> +  digitalWrite(LCD_PINS_RST  , LOW);
> +  delay(1000);
> +  digitalWrite(LCD_PINS_RST  , HIGH);
>
> Please indent, and please use FastIO:
>
>   SET_OUTPUT(LCD_PINS_DC);
>   SET_OUTPUT(LCD_PINS_RST);
>   WRITE(LCD_PINS_RST, LOW);
>   delay(1000);
>   WRITE(LCD_PINS_RST, HIGH);
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#pullrequestreview-67160281>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiRagj3dF3Ur9_8h_oVq4WBnSBfe8ks5so80cgaJpZM4PrIQI>
> .
>

jmdearras commented:
OK.  OEM supplied code patches.

On Wed, Oct 4, 2017 at 2:24 PM, Scott Lahteine <notifications@github.com>
wrote:

> *@thinkyhead* commented on this pull request.
> ------------------------------
>
> In Marlin/Marlin_main.cpp
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142752800>
> :
>
> >
> +
> +  /*---------------MKS OLED patch_3-----------------------*/
> +#if defined (MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
>
> Please remove /*---------------MKS OLED patch_3-----------------------*/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142752800>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiVdNy2JUafFCu94KrDmAfW6Wo79hks5so81QgaJpZM4PrIQI>
> .
>

jmdearras commented:
thanks for the input.  I will work on it.  Been programming since the early
70s!

On Wed, Oct 4, 2017 at 8:57 PM, Jim DeArras <jmdearras@gmail.com> wrote:

> OK.  OEM supplied code patches.
>
> On Wed, Oct 4, 2017 at 2:24 PM, Scott Lahteine <notifications@github.com>
> wrote:
>
>> *@thinkyhead* commented on this pull request.
>> ------------------------------
>>
>> In Marlin/Marlin_main.cpp
>> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142752800>
>> :
>>
>> >
>> +
>> +  /*---------------MKS OLED patch_3-----------------------*/
>> +#if defined (MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
>>
>> Please remove /*---------------MKS OLED patch_3-----------------------*/
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142752800>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAksiVdNy2JUafFCu94KrDmAfW6Wo79hks5so81QgaJpZM4PrIQI>
>> .
>>
>
>

jmdearras commented:
what is the intended difference between "if defined" (ifdef) and "if
enabled"

On Wed, Oct 4, 2017 at 2:03 PM, Scott Lahteine <notifications@github.com>
wrote:

> *@thinkyhead* commented on this pull request.
> ------------------------------
>
> In Marlin/pins_RAMPS.h
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r142746503>
> :
>
> > @@ -129,6 +129,21 @@
>    #define MAX6675_SS       66 // Do not use pin 49 as this is tied to the switch inside the SD card socket to detect if there is an SD card present
>  #endif
>
> +/*---------------MKS OLED patch_4-----------------------*/
> +#if defined (MKS_OLED13_128x64_FULL_GRAPHICS_CONTROLLER)
>
> #ifdef is preferred
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#pullrequestreview-67153837>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksidLdJwTMA8Dc_9MI4NEWWypK0DAJks5so8hogaJpZM4PrIQI>
> .
>

jmdearras committed:
Working on merge issues

jmdearras committed:
Merge branch 'Marlin-WIP' of https://github.com/jmdearras/Marlin-1 into Marlin-WIP

jmdearras committed:
more edits

jmdearras committed:
more cleanup, per requests

jmdearras committed:
cleanup

jmdearras commented:
Other than the description of the displays in configuration.h, have I caught all the issues?  I want to compile and test the displays before proceeding.

Jim

jmdearras committed:
Display name changes to match reprap

and descriptions added per the standard,

jmdearras committed:
sorry for the many commits

Just easier to see on the github site

jmdearras commented:
I believe I have addressed all the issues.  If the displays in configuration.h are acceptable, I will propagate them to the example configuration.h files.

jmdearras committed:
SanityCheck conflict fix

Oops!

jmdearras committed:
sanitycheck.h

added comment to display counter,

jmdearras committed:
comment correction

jmdearras committed:
Remove WIP comments

jmdearras commented:
Could I get some advice as to what the conflict is in Marlin/SanityCheck.h?

jmdearras committed:
working on sanitycheck.h conflict

jmdearras commented on Marlin/SanityCheck.h:
```diff
@@ -1146,13 +1147,16 @@ static_assert(1 >= 0
   #if ENABLED(ULTIMAKERCONTROLLER) \
       && DISABLED(SAV_3DGLCD) && DISABLED(miniVIKI) && DISABLED(VIKI2) \
-      && DISABLED(ELB_FULL_GRAPHIC_CONTROLLER) && DISABLED(PANEL_ONE)
+      && DISABLED(ELB_FULL_GRAPHIC_CONTROLLER) && DISABLED(PANEL_ONE) && DISABLED(MKS_12864OLED)
```
Why does this cause a conflict?

fiveangle replied:
You are adding it twice:

  #if ENABLED(ULTIMAKERCONTROLLER) \
      && DISABLED(SAV_3DGLCD) && DISABLED(miniVIKI) && DISABLED(VIKI2) \
      && DISABLED(ELB_FULL_GRAPHIC_CONTROLLER) && DISABLED(PANEL_ONE) && DISABLED(MKS_12864OLED)
    + 1
  #endif
  #if ENABLED(REPRAP_DISCOUNT_SMART_CONTROLLER) && DISABLED(REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER) &&  DISABLED(MKS_12864OLED)
    + 1
  #endif
  #if ENABLED(MKS_12864OLED)
     + 1

fiveangle replied:
```
* Note: BQ_LCD_SMART_CONTROLLER => REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
 *       REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER => REPRAP_DISCOUNT_SMART_CONTROLLER
 *       SAV_3DGLCD => U8GLIB_SH1106 => ULTIMAKERCONTROLLER
 *       MKS_12864OLED => U8GLIB_SH1106 => ULTIMAKERCONTROLLER
 *       miniVIKI => ULTIMAKERCONTROLLER
 *       VIKI2 => ULTIMAKERCONTROLLER
 *       ELB_FULL_GRAPHIC_CONTROLLER => ULTIMAKERCONTROLLER
 *       PANEL_ONE => ULTIMAKERCONTROLLER
```

Address the rest of the tidying up, if you're still struggling with this, I'll load this up and see if I can help,

jmdearras replied:
I beleive I have addressed all but the example configs, waiting to see if my configuration.h passes

fiveangle replied:
Are you looking at the site (and not just via email ?)  I have comments submitted that don't appear to be addressed.

jmdearras replied:
yes, now.

fiveangle commented:
Reviewing this, I had an idea... instead of creating new panel for esentially the identical panel except with the connector keys reversed, why not create a new FLIP_PANEL_CONNECTOR_ORIENTATION (or some other descriptive name) that will work for other panels too ?  How feasible is this (I haven't reveiwed the code) ?  With this feature, Minipanel would essentially become RRDFG with this feature set (and LCD contrast simply not connected).  

The reason I ask is that this is an issue that plagues many clone displays.  Most people struggle to implement, and then eventually find out that their panel has the keyways reversed, and so shave/re-install or remap the logical pins by hand.

With such a feature we would really improve the configuration for a lot of people, rather than adding panels for one orientation vs the other.  Thoughts ?

fiveangle commented on Marlin/Conditionals_LCD.h:
```diff
@@ -108,7 +108,28 @@
+  #elif ENABLED(CR10_STOCKDISPLAY)
+  
+  #define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
+  #ifndef ST7920_DELAY_1
```
[removed misinformation]
again - indention

jmdearras replied:
I was told I needed them.  :(

thinkyhead replied:
Now that they're wrapped in conditionals they are defaults — or "recommendations" — and they can be overridden in `Configuration.h` or `Configuration_adv.h`.

fiveangle commented on Marlin/Conditionals_LCD.h:
```diff
@@ -108,7 +108,28 @@
+    #define REPRAP_DISCOUNT_SMART_CONTROLLER 
+    #define NEWPANEL 
+
+  #elif ENABLED(MKS_MINI_12864)
```
Historically, users reverse the connectors.  This is not a problem isolated to MKS_MINI12864.  See my comments in the discussion regarding this.

jmdearras replied:
I am a EE, it is incorrect to install them backwards, defeats the reason for the key.  I believe it is a trick by the vendor to keep everything single sourced.

jmdearras replied:
comments removed

fiveangle commented on Marlin/SanityCheck.h:
```diff
@@ -1146,13 +1147,16 @@ static_assert(1 >= 0
-  #if ENABLED(REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER) && DISABLED(BQ_LCD_SMART_CONTROLLER)
+  #if ENABLED(MKS_12864OLED)
+     + 1
+   #endif
```
indention, indention, indention :)

fiveangle commented on Marlin/pins_RAMPS.h:
```diff
@@ -350,26 +379,43 @@
-      #define DOGLCD_CS         66
-      #define LCD_BACKLIGHT_PIN 65 // backlight LED on A11/D65
-      #define SDSS              53
+      #if ENABLED(MKS_MINI_12864)
```
If the pins are completely different, why nest here ?  if the intent here is to treat MKS_MINI12864 as different than MINIPANEL (still debatable) then treat it as a separate panel under it's own #elseif like every other, or it's too confusing.

Also - indention - it doesn't look like you un-indent this nest (even if you did).  Close your nests in the endif comment so it's not so difficult to follow:
`#endif // MINIPANEL`

jmdearras replied:
ok

fiveangle commented:
> the seller's page?

where someone can purchase the display.  but I gave you a pretty clear description of the comment fields up in the code comment that you may want to re-visit.

fiveangle commented on Marlin/Configuration.h:
```diff
@@ -1498,6 +1498,40 @@
+//
+//http://reprap.org/wiki/MKS_MINI_12864
+//
+// Be aware that MKS/makebase/makeboard install the 8 pin EXP1 and EXP2 headers reversed on all their products.  
```
Why mention this hear, since you are creating an entirely new panel entry for MINIPANEL that essentially flips the connectors logically ?  Doesn't this just add more confusion ?

jmdearras replied:
it does not flip power.

fiveangle commented on Marlin/Configuration.h:
```diff
@@ -1498,6 +1498,40 @@
+//
+// Factory display for Creality CR-10 
+//
+// this is a ramps-compatible version for CR-10 owners who want to replace the Creality board, but retain the display
```
Capitalization, product link.

fiveangle commented on Marlin/Configuration.h:
```diff
@@ -1498,6 +1498,40 @@
+// 
+//#define CR10_STOCKDISPLAY
+
+// MKS OLED 1.3" 128 × 64 FULL GRAPHICS CONTROLLER 
```
Missing blank comment leadin line

fiveangle commented on Marlin/Configuration.h:
```diff
@@ -1498,6 +1498,40 @@
+// MKS OLED 1.3" 128 × 64 FULL GRAPHICS CONTROLLER 
+//
+//http://reprap.org/wiki/MKS_12864OLED
+//
```
product link goes under description.  no extra blank lines w/o reason

jmdearras committed:
another try to fix conflict

jmdearras commented on Marlin/SanityCheck.h:
```diff
@@ -1146,13 +1147,16 @@ static_assert(1 >= 0
     + 1
   #endif
-  #if ENABLED(REPRAP_DISCOUNT_SMART_CONTROLLER) && DISABLED(REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER)
+  #if ENABLED(REPRAP_DISCOUNT_SMART_CONTROLLER) && DISABLED(REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER) && DISABLED(LCD_FOR_MELZI) && DISABLED(MKS_12864OLED)
```
This causes the conflict.  I do not see the problem.

jmdearras committed:
fix pins-ramps indents and if then.

jmdearras committed:
removed ifndef statements

The ones I was told to add earlier.  :-\

jmdearras committed:
removed comment

jmdearras committed:
display selections cleanup

jmdearras commented:
I am running into conflicts from comments.  A little frustrating.  But most advice has been sound and good.

jmdearras commented:
I will look at it, but it's not as simple as it seems.  Power and ground
are not reversed and define how the device must plug in. They are
hardwired, not assignable.

Jim

On Fri, Oct 6, 2017 at 10:16 AM, Dave Johnson <notifications@github.com>
wrote:

> Reviewing this, I had an idea... instead of creating new panel for
> esentially the identical panel except with the connector keys reversed, why
> not create a new FLIP_PANEL_CONNECTOR_ORIENTATION (or some other
> descriptive name) that will work for other panels too ? How feasible is
> this (I haven't reveiwed the code) ? With this feature, Minipanel would
> essentially become RRDFG with this feature set (and LCD contrast simply not
> connected).
>
> The reason I ask is that this is an issue that plagues many clone
> displays. Most people struggle to implement, and then eventually find out
> that their panel has the keyways reversed, and so shave/re-install or remap
> the logical pins by hand.
>
> With such a feature we would really improve the configuration for a lot of
> people, rather than adding panels for one orientation vs the other.
> Thoughts ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#issuecomment-334767567>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiW8eRmDCk7TtCNcaVgkYaD0BceQsks5spjY5gaJpZM4PrIQI>
> .
>

jmdearras commented:
it needs to be in both tests.  Or I need to remove the specific test for
the MKS_12864OLED.

Why can't I see this comment on the site, only email?

On Fri, Oct 6, 2017 at 10:25 AM, Dave Johnson <notifications@github.com>
wrote:

> *@fiveangle* commented on this pull request.
> ------------------------------
>
> In Marlin/SanityCheck.h
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r143203304>
> :
>
> > @@ -1146,13 +1147,16 @@ static_assert(1 >= 0
>  static_assert(1 >= 0
>    #if ENABLED(ULTIMAKERCONTROLLER) \
>        && DISABLED(SAV_3DGLCD) && DISABLED(miniVIKI) && DISABLED(VIKI2) \
> -      && DISABLED(ELB_FULL_GRAPHIC_CONTROLLER) && DISABLED(PANEL_ONE)
> +      && DISABLED(ELB_FULL_GRAPHIC_CONTROLLER) && DISABLED(PANEL_ONE) && DISABLED(MKS_12864OLED)
>
> You are adding it twice:
>
> #if ENABLED(ULTIMAKERCONTROLLER)
> && DISABLED(SAV_3DGLCD) && DISABLED(miniVIKI) && DISABLED(VIKI2)
> && DISABLED(ELB_FULL_GRAPHIC_CONTROLLER) && DISABLED(PANEL_ONE) &&
> DISABLED(MKS_12864OLED)
> + 1
> #endif
> #if ENABLED(REPRAP_DISCOUNT_SMART_CONTROLLER) &&
> DISABLED(REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER) &&
> DISABLED(MKS_12864OLED)
> + 1
> #endif
> #if ENABLED(MKS_12864OLED)
> + 1
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#discussion_r143203304>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiShsG0DaRsN4jvtg_Wm6t2QRyBSkks5spjhKgaJpZM4PrIQI>
> .
>

fiveangle commented:
> I will look at it, but it's not as simple as it seems.  Power and ground
> are not reversed and define how the device must plug in. They are
> hardwired, not assignable.

That's not what we've found.  MKS MINI12864 is essentially same pinout as RRDFGC with EXP1 and EXP2 each flipped 180 deg, and without support for some of the lcd brightness/contrast:

https://github.com/MarlinFirmware/Marlin/issues/6847#issuecomment-304587594

https://github.com/MarlinFirmware/Marlin/issues/6847#issuecomment-306229995

You can view the comments more easily [here on the website](https://github.com/MarlinFirmware/Marlin/pull/7824#) is all I was saying, as it seemed you missed many of my comments and updates here, so I was just trying to make sure technology isn't making communicating harder than it already is :)

jmdearras commented:
Can we look at the FLIP_PANEL_CONNECTOR_ORIENTATION later.  It still seems to me physical flipping is needed, software cannot fix the power pinout.  I just wanted to submit these displays I added for now.

jmdearras commented:
can you all confirm the configuration.h file is acceptable, so I can start the task of adding the changes to the example ones.?

jmdearras commented:
I will test the MKS MINI12864 against the RRDFGC tonight, I have both.  I have swapped the connector housing on all MKS stuff to put pin in the correct location.  The plastic housing can be easily pried up and reversed, avoiding the need to cut the key off the plug.

fiveangle commented on Marlin/Configuration.h:
```diff
@@ -1498,6 +1498,37 @@
+//
+// MKS Mini Panel with graphic controller and SD support 
+// http://reprap.org/wiki/MKS_MINI_12864
+//
```
remove the one extra blank line here

fiveangle commented on Marlin/Configuration.h:
```diff
@@ -1498,6 +1498,37 @@
+// Makeboard 3D Printer Parts 3D Printer Mini Display 1602 Mini Controller
+// https://www.aliexpress.com/item/Micromake-Makeboard-3D-Printer-Parts-3D-Printer-Mini-Display-1602-Mini-Controller-Compatible-with-Ramps-1/32765887917.html
+//
+// 2 line display.  
```
if you must add this info, add it to the description line:

`// Makeboard 3D Printer Parts 3D Printer Mini Display 1602 Mini Controller 2 line display`

and delete extra blank line

fiveangle commented on Marlin/Configuration.h:
```diff
@@ -1498,6 +1498,37 @@
+//#define MAKEBOARD_MINI_2_LINE_DISPLAY_1602
+
+//
+// MKS Mini Panel with graphic controller and SD support 
```
This panel is called MKS MINI12864 or MKS 12864LCD everywhere I see it.  Probably best to include the 12864 somehwere in the name so people will know what both this one and the OLED one are "12864"

jmdearras replied:
https://www.aliexpress.com/item/Micromake-Makeboard-3D-Printer-Parts-3D-Printer-Mini-Display-1602-Mini-Controller-Compatible-with-Ramps-1/32765887917.html

jmdearras replied:
or do you mean the mini panel?

fiveangle replied:
My comment is to line 1510:

`+// MKS Mini Panel with graphic controller and SD support `

fiveangle commented on Marlin/Configuration.h:
```diff
@@ -1498,6 +1498,37 @@
+// Factory display for Creality CR-10 
+// https://www.aliexpress.com/item/Universal-LCD-12864-3D-Printer-Display-Screen-With-Encoder-For-CR-10-CR-7-Model/32833148327.html
+//
+// this is a ramps-compatible version for CR-10 owners who want to replace the Melzi Creality board, but retain the display
```
Probably good to include the detail that this is "a RAMPS-compaitible, but with single 10-pin connector instead of 2" so people can more easily identify it (I certainly wasn't aware of that point - great of you to add it here).

thinkyhead commented:
> what is the intended difference between "if defined" (ifdef) and "if enabled"

Hey @jmdearras — welcome to the wonderful world of embedded programming, closest thing we now have to programming in the 70's-80's.

I put together a webpage to make sure coders are following the same standards. Here's a a link to the relevant section:

http://marlinfw.org/docs/development/coding_standards.html#macros

The short answer is that `ENABLED` is needed to allow setting a "disabled" state on `#define` config options using a `Makefile` or other external build method.

fiveangle commented:
And @jmdearras - don't be too frustrated - the LCD stuff in marlin is probably the most crazy of it all - definitely hard to follow sometimes !

Interested to see what you find out on your display pinouts when you check them.  I really hope they aren't yet another variant :)

I have the MKS MINI 12864LCD so can test on my config at the end to ensure it doesn't break.  Of course  I just use the existing MINIPANEL define currently, then have custom pin config in the pins_PRINTRBOARD_REVF.h since that board has different expansion ports than RAMPS, etc, but I can try to  adapt to use your define.

jmdearras committed:
more display cleanup

jmdearras commented:
I have always been an embedded programmer.  I designed and hold the patent on the first FedEx barcode reader, the boxy pencil one.  :)

jmdearras committed:
reverting a revertion.

jmdearras committed:
display description addition

thinkyhead committed:
Merge branch 'bugfix-1.1.x' into Marlin-WIP

thinkyhead commented:
@jmdearras Now that embedded has reached epic proportions, it's begun to feel a lot like the old **Homebrew Computer Club** days.

thinkyhead commented on Marlin/pins_RAMPS.h:
```diff
@@ -258,6 +264,23 @@
+      //#define SHIFT_OUT 40
+      //#define SHIFT_EN 17
+    #endif
+	#else
```
No TABS — Indentation is **two spaces**. Please change your editor settings.

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -1503,6 +1503,37 @@
+//
+// this is a a RAMPS-compaitible version, but with a single 10-pin connector instead of 2, for CR-10 owners who want to replace the Melzi Creality board, but retain the display
+// 
+//#define CR10_STOCKDISPLAY
```
Let's see if I understand correctly. This is the option someone would select using the CR-10 display on a RAMPS-compatible board, and since the CR-10 sees it as a standard Melzi plus RRD Full Graphical Smart Display, this option may also be selected for this display in a CR-10 — since it just forces `REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER` to be enabled.

I think I have enough to work with from this PR to clean it up and merge it. I will follow up with a new PR soon.

jmdearras replied:
yes

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -13583,8 +13583,14 @@ void setup() {
+  SET_OUTPUT(LCD_PINS_RST);  
+  WRITE(LCD_PINS_RST, LOW);
+  delay(1000);
+  WRITE(LCD_PINS_RST, HIGH);
```
Is this supposed to be a different pin from `LCD_PINS_RS`? Your PR is the only place this pin appears.

jmdearras replied:
patch file from MKS had it,

thinkyhead replied:
Can you point me to the author of that patch? I can redirect some of these questions to them.

jmdearras replied:
I commented out that whole reset section and renamed the pin to LCD_PINS_RS in pins_ramps.h, and it does not seem to have affected the display, it still works.  The code was copied from the OEM recommendation, probably for an earlier marlin release.

thinkyhead replied:
It's an interesting piece. If it does no harm, I guess it can be left in. Anyway, off to figure out what's wrong with MKS_12864OLED.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -13583,8 +13583,14 @@ void setup() {
-}
-
+#if defined (MKS_12864OLED)
+  SET_OUTPUT(LCD_PINS_DC);   
```
Also, this is the only place this pin is used. Does this `SET_OUTPUT` do something special that the display needs, even though the pin is never used after this?

jmdearras replied:
It was in the patch file for the display.  I can remove it and see if it still works.

thinkyhead replied:
This pin coincides with `DOGLCD_A0` — they are both assigned to pin 25. During init I guess this pin takes on a different role…?

thinkyhead commented on Marlin/ultralcd_impl_DOGM.h:
```diff
@@ -183,9 +183,13 @@
   U8GLIB_SSD1306_128X64_2X u8g(U8G_I2C_OPT_NONE | U8G_I2C_OPT_FAST); // 4 stripes
 #elif ENABLED(U8GLIB_SH1106)
+  #if ENABLED(MKS_12864OLED)
+    U8GLIB_SH1106_128X64 u8g(23, 17, 16, 25); // SW SPI Com: SCK = 23, MOSI = 17, CS = 16, A0 = 25
```
What's with the hard-coded pin numbers? According to these additions to the `pins_RAMPS.h` file, we end up with these pins:
```cpp
    #if ENABLED(MKS_12864OLED)
      #define LCD_PINS_DC  25
      #define LCD_PINS_RST 27
    #else
      #define LCD_PINS_D5  25
      #define LCD_PINS_D6  27
    #endif

    #define LCD_PINS_RS    16
    #define LCD_PINS_ENABLE 17
    #define LCD_PINS_D4    23
    #define LCD_PINS_D7    29
```
So this line might be written instead as:
```cpp
U8GLIB_SH1106_128X64 u8g(LCD_PINS_D4, LCD_PINS_ENABLE, LCD_PINS_RS, LCD_PINS_DC);
```
Was that your intention?

jmdearras replied:
I am not familiar with your functions and macros.  That was the OEM recommendation.  If you show me the final code, I'll try it.

jmdearras replied:
ok, i understand, now.  I'll try it

jmdearras replied:
that worked

jmdearras committed:
MKS_12864OLED cleanup

Works  Not sure about the reset I removed in marlin_main, but the display seems to work fine without it.

jmdearras committed:
Merge branch 'Marlin-WIP' of https://github.com/jmdearras/Marlin-1 into Marlin-WIP

jmdearras commented:
hold off on commit, issue with MKS_MINI_12864 after changing the physical order of things,

jmdearras commented:
quitting for the night, I am suddenly getting errors on ENCODER_PULSES_PER_STEP and PER_MENU_ITEM, no clue how I broke them.

jmdearras committed:
encoder stuff??

jmdearras commented:
I closed it by mistake

thinkyhead commented on Marlin/ultralcd_impl_DOGM.h:
```diff
@@ -183,9 +183,14 @@
 #elif ENABLED(U8GLIB_SH1106)
+  #if ENABLED(MKS_12864OLED)
+ //   U8GLIB_SH1106_128X64 u8g(23, 17, 16, 25); // SW SPI Com: SCK = 23, MOSI = 17, CS = 16, A0 = 25
+    U8GLIB_SH1106_128X64 u8g(LCD_PINS_D4, LCD_PINS_ENABLE, LCD_PINS_RS, LCD_PINS_DC);
```
Changed to…
```cpp
  // MKS 128x64 (SH1106) OLED I2C LCD
  U8GLIB_SH1106_128X64 u8g(DOGLCD_SCK, DOGLCD_MOSI, DOGLCD_CS, DOGLCD_A0);
```

thinkyhead commented:
#7867 should now have the right `NEWPANEL` pins — I had forgotten to set `REPRAP_DISCOUNT_SMART_CONTROLLER`. If that passes, it might get into 1.1.6. Just putting release notes together now….

jmdearras commented:
edited `REPRAP_DISCOUNT_SMART_CONTROLLER`. in sanity check
```cpp
  #if ENABLED(REPRAP_DISCOUNT_SMART_CONTROLLER) \
      && DISABLED(REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER) \
      && DISABLED(LCD_FOR_MELZI) \
      && DISABLED(MAKEBOARD_MINI_2_LINE_DISPLAY_1602)
      && DISABLED(MKS_12864OLED)
    + 1
  #endif
```
clean compile and it works.

Jim


On Mon, Oct 9, 2017 at 5:36 AM, Scott Lahteine <notifications@github.com>
wrote:

> #7867 <https://github.com/MarlinFirmware/Marlin/pull/7867> should now
> have the right NEWPANEL pins — I forgot to set REPRAP_DISCOUNT_SMART_
> CONTROLLER.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#issuecomment-335108224>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiagsD7Cokm-PpxtepyepFFNtfU2qks5sqekEgaJpZM4PrIQI>
> .
>

jmdearras committed:
update all

ScruffR commented:
Sorry for commenting here but I just couldn't make sense of what I found on the net so far and this discussion did lighten things up a bit, but it also added some confusion 😳 

I'm trying to use the stock CR-10 display with a GT2560 board to add dual extrusion to my CR-10, so I tried 1.1.6 with
```
#define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
#define U8GLIB_SH1106
#define CR10_STOCKDISPLAY
```
but I only get short blink and a chirp from the buzzer when switching on the board.
Do I need to rotate the 10pin header so that top left pin ends up bottom right? 
Or do I need to flip the pins so that top left ends up top right? 
Or both so that top left ends up bottom left?

jmdearras commented:
if it lights up, it's correct  otherwise, try reversing it.

you only need to define CR10_STOCKDISPLAY.

Jim

On Tue, Oct 17, 2017 at 2:44 PM, ScruffR <notifications@github.com> wrote:

> Sorry for commenting here but I just couldn't make sense of what I found
> on the net so far and this discussion did lighten things up a bit, but it
> also added some confusion 😳
>
> I'm trying to use the stock CR-10 display with a GT2560 board to add dual
> extrusion to my CR-10, so I tried 1.1.6 with
>
> #define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
> #define U8GLIB_SH1106
> #define CR10_STOCKDISPLAY
>
> but I only get short blink and a chirp from the buzzer when switching on
> the board.
> Do I need to rotate the 10pin header so that top left pin ends up bottom
> right?
> Or do I need to flip the pins so that top left ends up top right?
> Or both so that top left ends up bottom left?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#issuecomment-337329490>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksicrkRq7hgn4VL_Kc9RwpG7GNNMamks5stPWbgaJpZM4PrIQI>
> .
>

ScruffR commented:
Thanks!
I'll comment `REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER` and try rotating the connector (after shaving off the polarity protection) then.

jmdearras commented:
OK.  I do not know why some board makers reverse the aux connectors, the
CR-10 stock one is correct.


On Wed, Oct 18, 2017 at 2:07 AM, ScruffR <notifications@github.com> wrote:

> Thanks!
> I'll comment REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER and try
> rotating the connector (after shaving off the polarity protection) then.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#issuecomment-337469774>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiXMgbZXGLlUbXYiJ0sytfSKRUU0zks5stZWugaJpZM4PrIQI>
> .
>

AnHardt commented:
I do know.
It's this ingenious pdf where you cant say what's on the right and what's on the left.
http://reprap.org/mediawiki/images/7/79/LCD_connect_SCHDOC.pdf

jmdearras commented:
pin 1 is defined by the maker of the socket. on the key side of the
connector.


On Wed, Oct 18, 2017 at 1:28 PM, AnHardt <notifications@github.com> wrote:

> I do know.
> It's this ingenious pdf where you cant say what's on the right and what's
> on the left.
> http://reprap.org/mediawiki/images/7/79/LCD_connect_SCHDOC.pdf
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7824#issuecomment-337665976>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAksiWQOZl-Zh7c8uN4OeBjcKfc764bWks5stjVFgaJpZM4PrIQI>
> .
>

jmdearras commented:
[image: Inline image 1]

On Wed, Oct 18, 2017 at 1:39 PM, Jim DeArras <jmdearras@gmail.com> wrote:

> pin 1 is defined by the maker of the socket. on the key side of the
> connector.
>
>
> On Wed, Oct 18, 2017 at 1:28 PM, AnHardt <notifications@github.com> wrote:
>
>> I do know.
>> It's this ingenious pdf where you cant say what's on the right and what's
>> on the left.
>> http://reprap.org/mediawiki/images/7/79/LCD_connect_SCHDOC.pdf
>>
>> —
>> You are receiving this because you modified the open/close state.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/MarlinFirmware/Marlin/pull/7824#issuecomment-337665976>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAksiWQOZl-Zh7c8uN4OeBjcKfc764bWks5stjVFgaJpZM4PrIQI>
>> .
>>
>
>

ScruffR commented:
I've now tried with the connector rotated and now the display lights up and the buzzer chirps when keeping the encoder button pressed, but the display stays blank

Tried all combinations of 
```
//#define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
#define U8GLIB_SH1106
#define CR10_STOCKDISPLAY
```
and
```
#define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
#define U8GLIB_SH1106
#define CR10_STOCKDISPLAY
```
and
```
BOARD_GT2560_REV_A 
```
and
```
BOARD_GT2560_REV_A_PLUS
```
(it's a board marked as A+)

The display is connected to the 10 pin header marked LCD (I guess this is EXT2 - not AUX)

BTW, gcode M300 does play a sound as intended.

These are my configs for 1.1.6


[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/1396269/Configuration.zip)

ScruffR commented:
I have now flashed some test sketch onty my GT2560 A+ board and realised that the pinout on the header must be completely off.

With a u8glib test sketch I figured that the CR10_STOCKDISPLAY actually uses only three pins (SPI mode).
On the GT2560 A+ the respective pins are 
```
// CR10_STOCKDISPLAY on GT2560 | name on DISPLAY | pin# on GT2560 | function
#define E  21                 // E               | 21             | SCLK  (SPI Clock) 
#define RW  6                 // RW              |  6             | MOSI  (Master Out Slave In)
#define RS  5                 // RS              |  5             | CS/SS (Chip/Slave Select)
```
_(this is with the 10pin header rotated)_

ScruffR commented:
I haven't come round yet to investigate the SD pinout but this is how the display and rotary encoder need to be set up for CR10_STOCKDISPLAY on GT2560 A/A+

`pins_GT2560_REV_A.h`
```
#if ENABLED(ULTRA_LCD)

  #define BEEPER_PIN       18
  #if ENABLED(NEWPANEL)
    // ************* ScruffR ************* 
    #if ENABLED(CR10_STOCKDISPLAY) 
      #define LCD_PINS_RS      5
      #define LCD_PINS_ENABLE  6
      #define LCD_PINS_D4     21
      #define LCD_PINS_D5     -1
      #define LCD_PINS_D6     -1
      #define LCD_PINS_D7     -1

      // Buttons are directly attached
      #define BTN_EN1         17
      #define BTN_EN2         16
      #define BTN_ENC         19
    // ************* ScruffR ************* 
    #else
      #define LCD_PINS_RS     20
      #define LCD_PINS_ENABLE 17
      #define LCD_PINS_D4     16
      #define LCD_PINS_D5     21
      #define LCD_PINS_D6      5
      #define LCD_PINS_D7      6

      // Buttons are directly attached
      #define BTN_EN1         42
      #define BTN_EN2         40
      #define BTN_ENC         19
    #endif

    #define SD_DETECT_PIN     38

  #else // !NEWPANEL

    #define SHIFT_CLK      38
    #define SHIFT_LD       42
    #define SHIFT_OUT      40
    #define SHIFT_EN       17

    #define LCD_PINS_RS      5
    #define LCD_PINS_ENABLE  6
    #define LCD_PINS_D4     21
    #define LCD_PINS_D5      0
    #define LCD_PINS_D6      0
    #define LCD_PINS_D7      0

    #define SD_DETECT_PIN  -1

  #endif // !NEWPANEL

#endif // ULTRA_LCD
```
